### DISCUSSION TITLE
duckdb in AWS-Lambda: Memory/Storage issues

### DISCUSSION THREAD

**Discussion by bhofner** [duckdb/duckdb#8687] [Q&A]
Hello,

I created an AWS Lambda function using the DuckDB Python v.0.8.1.
The Lambda has 10 GB memory and 10 GB /tmp space.
Since that was not enough, I mounted an AWS EFS volume.
Furthermore, the HTTPFS-Extension for Linux ist installed (linux_amd64_gcc4)

DuckDB-Settings are: 
```
SET memory_limit='8GB';
SET home_directory='/tmp/duckdb'
SET temp_directory='/tmp/duckdb-tmp'
```

This generally works well.

It gets problematic if the ingested CSV gets too big (~ 10GB uncompressed size, ~46 mio rows).

Symptoms:
- Writing the ingested table to S3 in parquet format times out.
- OutOfMemory errors when performing grouping queries.

I tried to changed the DuckDB temp directory to my EFS mount, which is probably much slower than the local /tmp:
`SET temp_directory='/mnt/efs/duckdb'
`
That produces sometimes:

- SegmentationFaults or
- IOException: "what() IOError could not truncate file "/mnt/efs/duckdb/duckdb_tem_storage-0.tmp: Stale file handle" or
- IOException: "IOError Could not read all bytes from file "/mnt/efs/duckdb/duckdb_tem_storage-0.tmp: wanted: 262144 read: 0" or

I switched back to the local /tmp folder, skipped some of my queries (e.g. group by's looking for duplicates) and
switched exporting from s3 to the efs volume.
That works for now, but seems right on the edge of what is possible in this environment.

Is there a way around these memory / temp space limitations?
E.g. a more reliable way to use AWS EFS as tmp space?

Best regards,

Bernd

**Comment by bhofner** [comment]
I concluded for me, that AWS Lambda is currently not a good environment for running DuckDB - at least if the tables to be processed won't fit for sure in the RAM provided by the Lambda.
Nevertheless there might be potential to improve the behaviour of DuckDB if memory gets tight.

**Reply by adiii717** [reply]
We hit the same issue recently with a 15M row / 87 column table timing out on dedup in Lambda.
 
The fix was replacing `DISTINCT ON` with a two-step `ROWID` approach:
 
**Step 1: Hash only 2 columns**
 
```sql
CREATE TABLE winners AS
SELECT key_column, MIN(rowid) AS rid
FROM big_table
GROUP BY key_column;
```
 
**Step 2: Join to get full rows**
 
```sql
CREATE TABLE deduped AS
SELECT t.*
FROM big_table t
JOIN winners w ON t.rowid = w.rid;
```
 
This reduced memory from 10GB (crash) to 5.5GB and time from timeout to 49 seconds.
